File: ios/Plugin/AudioFile.swift
//
//  AudioFile.swift
//  Plugin
//
//  Created by Skylabs Technology on 5/20/21.
//  Copyright © 2021 Max Lynch. All rights reserved.
//

import Foundation
import AVFoundation

public class AudioFile {
    public let engine: AVAudioEngine
    public let player: AVAudioPlayerNode = AVAudioPlayerNode()
    public let eq: AVAudioUnitEQ = AVAudioUnitEQ(numberOfBands: 3)
    public var _parent: Mixer
    public let playerMixer: AVAudioMixerNode = AVAudioMixerNode()
    public var playerQueue: DispatchQueue
    public var meterQueue: DispatchQueue
    
    public var audioFile: AVAudioFile?
    public var audioSampleRate: Double = 0
    public var audioLengthSeconds: Double = 0
    public var audioLengthSamples: AVAudioFramePosition = 0
    public var needsFileScheduled = true
    public var seekFrame: AVAudioFramePosition = 0
    public var listenerName: String = ""
    
    public var elapsedTime: TimeInterval = 0
    public var elapsedTimeEventName: String = ""
    
    init(parent: Mixer, audioId: String){
        _parent = parent
        engine = _parent.engine
        playerQueue = DispatchQueue(label: "mixerPlugin.audioFile.queue.\(audioId)", qos: .userInitiated)
        meterQueue = DispatchQueue(label: "mixerPlugin.audioMeter.queue.\(audioId)", qos: .userInitiated)
    }
    
    /**
     * Starts initialization of a player. Configures player then starts its listeners
     *
     * @param audioFilePath
     * @param channelSettings
     */
    // MARK: setupAudio
    public func setupAudio(audioFilePath: NSURL, channelSettings: ChannelSettings) {
      do {
        let file = try AVAudioFile(forReading: audioFilePath as URL)
        let format = file.processingFormat
//        let newEqSettings = channelSettings.eqSettings
        
        audioLengthSamples = file.length
        audioSampleRate = format.sampleRate
        audioLengthSeconds = Double(audioLengthSamples) / audioSampleRate
        audioFile = file
        
        setupEq(with: format, channelSettings: channelSettings)
      } catch {
        // print("Error reading the audio file: \(error.localizedDescription)")
      }
    }
    
    /**
     * Sets up EQ and attaches it to player
     *
     * @param channelSettings
     */
    // MARK: setupEq
    private func setupEq(with format: AVAudioFormat, channelSettings: ChannelSettings) {
        let bassEq = eq.bands[0]
        bassEq.filterType = .lowShelf
        bassEq.frequency = channelSettings.eqSettings!.bassFrequency!
        bassEq.gain = channelSettings.eqSettings!.bassGain!
        bassEq.bypass = false
        
        let midEq = eq.bands[1]
        midEq.filterType = .parametric
        midEq.frequency = channelSettings.eqSettings!.midFrequency!
        midEq.bandwidth = 1
        midEq.gain = channelSettings.eqSettings!.midGain!
        midEq.bypass = false
        
        let trebleEq = eq.bands[2]
        trebleEq.filterType = .highShelf
        trebleEq.frequency = channelSettings.eqSettings!.trebleFrequency!
        trebleEq.gain = channelSettings.eqSettings!.trebleGain!
        trebleEq.bypass = false
        
        configureEngine(with: format, channelSettings: channelSettings)
    }
    
    /**
     * Completes remaing setup for player and enables EQ
     *
     * @param format
     * @param channelSettings
     */
    // MARK: configureEngine
    private func configureEngine(with format: AVAudioFormat, channelSettings: ChannelSettings) {
        if (engine.isRunning) {
            engine.stop()
        }
        player.volume = 0.8
        playerMixer.outputVolume = channelSettings.volume!
        
        engine.attach(player)
        engine.attach(eq)
        engine.attach(playerMixer)
        
        if (channelSettings.channelListenerName != "") {
            listenerName = channelSettings.channelListenerName!
            playerMixer.removeTap(onBus: 0)
            playerMixer.installTap(onBus: 0, bufferSize: 1024, format: playerMixer.outputFormat(forBus: 0), block: handleMetering)
        }

        if (channelSettings.elapsedTimeEventName != "") {
            setElapsedTimeEvent(eventName: channelSettings.elapsedTimeEventName!)
        }
        
        engine.connect(player, to: eq, format: format)
        engine.connect(eq, to: playerMixer, format: format)
        engine.connect(playerMixer, to: engine.mainMixerNode, format: playerMixer.outputFormat(forBus: 0))
        engine.prepare()
        
      
      do {
        try engine.start()
        
        scheduleAudioFile()
//        isPlayerReady = true
      } catch {
        // print("Error starting the player: \(error.localizedDescription)")
      }
    }
    
    /**
     * Sets up an audio file when initilized or needs to be reset
     */
    // MARK: scheduleAudioFile
    public func scheduleAudioFile() {
        guard let file = audioFile, needsFileScheduled else {
            return
        }
        needsFileScheduled = false
        seekFrame = 0
        player.volume = 0.8
        player.scheduleFile(file, at: nil) {
            self.needsFileScheduled = true
            DispatchQueue.main.asyncAfter(deadline: .now()+1.0) {
                self.player.stop()
            }
        }
        // playOrPause()
    }
    
    /**
     * Local listener that handles metering 
     * 
     * @param buffer
     * @param time
     */
    // MARK: handleMetering
    public func handleMetering(buffer: AVAudioPCMBuffer, time: AVAudioTime) {
        // https://www.raywenderlich.com/21672160-avaudioengine-tutorial-for-ios-getting-started
        if (player.isPlaying) {
            meterQueue.async {
                guard let channelData = buffer.floatChannelData else { return }
                  
                let channelDataValue = channelData.pointee

                let channelDataValueArray = stride(
                  from: 0,
                  to: Int(buffer.frameLength),
                  by: buffer.stride)
                  .map { channelDataValue[$0] }
                  
                let rms = sqrt(channelDataValueArray.map {
                    return $0 * $0
                }
                .reduce(0, +) / Float(buffer.frameLength))
                  
                let avgPower = 20 * log10(rms)
        //          let meterLevel = self.scaledPower(power: avgPower)
                let response = avgPower < -80 ? -80 : avgPower

                self._parent.notifyListeners(self.listenerName, data: [ResponseParameters.meterLevel: response])
            
                if (self.elapsedTimeEventName != "") {
                    self._parent.notifyListeners(self.elapsedTimeEventName, data: self.player.elapsedTimeSeconds.toDictionary())
                }
            }
        }
    }
    
    /**
     * Adds the elapsedTimeEventName and enables playback notifications for player elapsed time
     *
     * @param eventName
     */
    public func setElapsedTimeEvent(eventName: String) {
        elapsedTimeEventName = eventName
    }
    

    // MARK: Player Controls
    
    /**
     * Handles play or pause for player
     *
     * @return
     */
    // MARK: playOrPause
    public func playOrPause() -> String {
        if (self.player.isPlaying) {
            self.player.pause()
            return "pause"
        } else {
            if (self.needsFileScheduled) {
                self.scheduleAudioFile()
            }
            playerQueue.async {
                self.player.play()
            }
            return "play"
        }
    }
    
    /**
     * Handles "stop" for player
     *
     * @return
     */
    // MARK: stop
    public func stop() -> String {
        elapsedTime = 0
        player.stop()
        return "stop"
//        needsFileScheduled = true
    }
    
    /**
     * Returns player if it is playing
     *
     * @return
     */
    // MARK: isPlaying
    public func isPlaying() -> Bool {
        return player.isPlaying
    }
    
    /**
     * Changes volume for the player
     *
     * @param volume
     */
    // MARK: adjustVolume
    public func adjustVolume(volume: Float) {
        playerMixer.outputVolume = volume
    }
    
    /**
     * Returns current volume for player
     *
     * @return
     */
    // MARK: getCurrentVolume
    public func getCurrentVolume() -> Float {
        return playerMixer.outputVolume
    }
    
    /**
     * Changes EQ output associated with the player
     *
     * @param type
     * @param gain
     * @param freq
     */
    // MARK: adjustEq
    public func adjustEq(type: String, gain: Float, freq: Float) {
        if(eq.bands.count < 1) {
            return
        }
        switch type.lowercased() {
        case "bass":
            let bassEq = eq.bands[0]
            bassEq.gain = gain
            bassEq.frequency = freq.isEqual(to: -1) ? bassEq.frequency : freq
            break;
        case "mid":
            let midEq = eq.bands[1]
            midEq.gain = gain
            midEq.frequency = freq.isEqual(to: -1) ? midEq.frequency : freq
            break;
        case "treble":
            let trebleEq = eq.bands[2]
            trebleEq.gain = gain
            trebleEq.frequency = freq.isEqual(to: -1) ? trebleEq.frequency : freq
            break;
        default: break
            // print("adjustEq: invalid eq type")
        }
    }
    
    /**
     * Returns current tracked EQ
     * @return
     */
    // MARK: getCurrentEq
    public func getCurrentEq() -> [String: Float] {
        if(eq.bands.count < 1) {
            return [:]
        }
        let bassEq = eq.bands[0]
        let midEq = eq.bands[1]
        let trebleEq = eq.bands[2]
        
        return [ResponseParameters.bassGain: bassEq.gain,
                ResponseParameters.bassFrequency: bassEq.frequency,
                ResponseParameters.midGain: midEq.gain,
                ResponseParameters.midFrequency: midEq.frequency,
                ResponseParameters.trebleGain: trebleEq.gain,
                ResponseParameters.trebleFrequency: trebleEq.frequency]
    }
    
    /**
     * Returns current elapsed time on player
     *
     * Note: This can be done automatically using setElapsedTimeEvent
     *
     * @return
     */
    // MARK: getElapsedTime
    public func getElapsedTime() -> [String: Int] {
        if(player.elapsedTimeSeconds > elapsedTime) {
            elapsedTime = player.elapsedTimeSeconds
        }
        return elapsedTime.toDictionary()
    }
    
    /**
     * Returns total time for the loaded track
     * @return
     */
    // MARK: getTotalTime
    public func getTotalTime() -> [String: Int] {
        return (audioFile?.totalDurationSeconds.toDictionary())!
    }
    
    /**
     * Destroys object and resets state.
     *
     * @return
     */
    // MARK: Destroy
    public func destroy() -> [String : String] {
        if(player.isPlaying){
            player.stop()
        }
        playerMixer.removeTap(onBus: 0)
        
        engine.detach(player)
        engine.detach(eq)
        engine.detach(playerMixer)
        return [ResponseParameters.listenerName: self.listenerName,
                ResponseParameters.elapsedTimeEventName: self.elapsedTimeEventName]
    }
}

extension TimeInterval{

    /**
     * Converts TimeInterval to a dictionary that has seperated to milliSeconds, seconds, minutes and hours
     */
    func toDictionary() -> [String : Int] {
        let time = NSInteger(self)
        let milliSeconds = Int((self.truncatingRemainder(dividingBy: 1)) * 1000)
        let seconds = time % 60
        let minutes = (time / 60) % 60
        let hours = (time / 3600)

        return [
            ResponseParameters.milliSeconds: milliSeconds,
            ResponseParameters.seconds: seconds,
            ResponseParameters.minutes: minutes,
            ResponseParameters.hours: hours
        ]
    }
}

extension AVAudioFile {

    /**
     * Returns the total duration of the AudioFile
     */
    var totalDurationSeconds: TimeInterval {
        let sampleRateSong = Double(processingFormat.sampleRate)
        let lengthSongSeconds = Double(length) / sampleRateSong
        return lengthSongSeconds
    }
}

extension AVAudioPlayerNode {
    
    private struct currentelapsed {
        static var time: TimeInterval = 0
    }
    
    /**
     * Returns the current elapsed time of the AudioFile
     */
    var currentElapsedTime: TimeInterval {
        get {
            guard let time = objc_getAssociatedObject(self, &currentelapsed.time) as? TimeInterval else {
                return 0
            }
            return time
        }
        set {
            objc_setAssociatedObject(self, &currentelapsed.time, newValue, .OBJC_ASSOCIATION_RETAIN)
        }
    }

    /**
     * Returns the current elapsed time in seconds of the AudioFile
     */
    var elapsedTimeSeconds: TimeInterval {
        if let nodeTime = lastRenderTime, let playerTime = playerTime(forNodeTime: nodeTime) {
            return Double(playerTime.sampleTime) / playerTime.sampleRate
        }
        return 0
    }
    
}

File: android/src/main/java/com/skylabs/mixer/AudioFile.java
package com.skylabs.mixer;
import android.content.res.AssetFileDescriptor;
import android.media.AudioAttributes;
import android.media.AudioManager;
import android.media.MediaPlayer;
import android.media.audiofx.DynamicsProcessing;
import android.media.audiofx.Visualizer;
import android.net.Uri;
import android.os.ParcelFileDescriptor;
import android.util.Log;
import android.media.audiofx.DynamicsProcessing.Eq;
import android.media.audiofx.DynamicsProcessing.EqBand;


import com.getcapacitor.JSObject;

import java.io.File;
import java.util.HashMap;
import java.util.Map;

import static com.skylabs.mixer.Utils.getPath;

public class AudioFile implements MediaPlayer.OnPreparedListener, MediaPlayer.OnCompletionListener {
    private Mixer _parent;
    private MediaPlayer player;
    private Eq eq;
    private DynamicsProcessing dp;
    private float currentVolume;
    public String elapsedTimeEventName = "";
    public String listenerName = "";

    private Visualizer visualizer;
    private boolean visualizerState = false;
    private Visualizer.MeasurementPeakRms measurementPeakRms;


    public AudioFile(Mixer parent) {
        _parent = parent;
        player = new MediaPlayer();
        player.setOnCompletionListener(this);
        player.setOnPreparedListener(this);
    }

    /**
     * Starts initialization of an MediaPlayer. Configures player then starts its listeners
     *
     * @param audioFilePath
     * @param channelSettings
     */
    public void setupAudio(String audioFilePath, ChannelSettings channelSettings) {
        try {
            Uri uri = Uri.parse(audioFilePath);
            String filePath = getPath(_parent._context, uri);
            File file = new File(filePath);
            ParcelFileDescriptor pfd = ParcelFileDescriptor.open(file, ParcelFileDescriptor.MODE_READ_ONLY);
            AssetFileDescriptor afd = new AssetFileDescriptor(pfd, 0, -1);
            player.setAudioAttributes(new AudioAttributes.Builder()
                            .setUsage(AudioAttributes.USAGE_MEDIA)
                            .setContentType(AudioAttributes.CONTENT_TYPE_MUSIC)
                            .setLegacyStreamType(AudioManager.STREAM_MUSIC)
                            .setFlags(AudioAttributes.FLAG_AUDIBILITY_ENFORCED)
                            .build()
            );
            player.setDataSource(afd.getFileDescriptor(), afd.getStartOffset(), afd.getLength());
            player.prepare();
            setupEq(channelSettings);
        }
        catch(Exception ex) {
            Log.e("setupAudio", "Exception thrown in setupAudio: " + ex);
        }
    }

    /**
     * Sets up EQ and attaches it to MediaPlayer
     *
     * @param channelSettings
     */
    private void setupEq(ChannelSettings channelSettings) {
        EqBand bassEq = new EqBand(true, (float)channelSettings.eqSettings.bassFrequency, (float)channelSettings.eqSettings.bassGain);
        EqBand midEq = new EqBand(true, (float)channelSettings.eqSettings.midFrequency, (float)channelSettings.eqSettings.midGain);
        EqBand trebleEq = new EqBand(true, (float)channelSettings.eqSettings.trebleFrequency, (float)channelSettings.eqSettings.trebleGain);
        eq = new Eq(true, true, 3);
        eq.setBand(0, bassEq);
        eq.setBand(1, midEq);
        eq.setBand(2, trebleEq);
        DynamicsProcessing.Config config = new DynamicsProcessing.Config.Builder(
                DynamicsProcessing.VARIANT_FAVOR_FREQUENCY_RESOLUTION,
                1,
                false, 0,
                false, 0,
                true, 3,
                false
        ).setPreferredFrameDuration(10).build();
        dp = new DynamicsProcessing(0, player.getAudioSessionId(), config);
        dp.setPostEqAllChannelsTo(eq);
        configureEngine(channelSettings);
    }

    /**
     * Completes remaing setup for MediaPlayer and enables EQ
     *
     * @param channelSettings
     */
    private void configureEngine(ChannelSettings channelSettings) {
        if (!channelSettings.channelListenerName.isEmpty()) {
            listenerName = channelSettings.channelListenerName;
        }
        if (!channelSettings.elapsedTimeEventName.isEmpty()) {
            setElapsedTimeEvent(channelSettings.elapsedTimeEventName);
        }
        dp.setEnabled(true);
        currentVolume = (float)channelSettings.volume;
        player.setVolume(currentVolume, currentVolume);
    }

    /**
     * Adds the elapsedTimeEventName and enables playback notifications for MediaPlayer elapsed time
     *
     * @param eventName
     */
    public void setElapsedTimeEvent(String eventName) {
        elapsedTimeEventName = eventName;
    }

    /**
     * Handles play or pause for MediaPlayer
     *
     * @return
     */
    public String playOrPause() {
        if (player.isPlaying()) {
            player.pause();
            if(visualizerState) {
                destroyVisualizerListener();
            }
            return "pause";
        } else {
            player.start();
            if(!visualizerState) {
                initVisualizerListener();
            }
            return "play";
        }
    }

    /**
     * Handles "stop" for MediaPlayer
     *
     * @return
     */
    public String stop() {
        if (player.isPlaying()) {
            player.pause();
        }
        player.seekTo(0);
        if(visualizerState) {
            destroyVisualizerListener();
        }
        return "stop";
    }

    /**
     * Returns MediaPlayer if it is playing
     *
     * @return
     */
    public boolean isPlaying() {
        return player.isPlaying();
    }

    /**
     * Changes volume for the MediaPlayer
     *
     * @param volume
     */
    public void adjustVolume(double volume) {
        currentVolume = (float)volume;
        player.setVolume(currentVolume, currentVolume);
    }

    /**
     * Returns current volume for MediaPlayer
     *
     * @return
     */
    public double getCurrentVolume() {
        return currentVolume;
    }

    /**
     * Changes EQ output associated with the MediaPlayer
     *
     * @param type
     * @param gain
     * @param freq
     */
    public void adjustEq(String type, double gain, double freq) {
        if (eq.getBandCount() < 1) {
            return;
        }
        if (type.equals("bass")) {
            EqBand bassEq = eq.getBand(0);
            bassEq.setGain((float) gain);
            bassEq.setCutoffFrequency((float) freq);
            eq.setBand(0, bassEq);
            dp.setPostEqAllChannelsTo(eq);
        }
        else if (type.equals("mid")) {
            EqBand midEq = eq.getBand(1);
            midEq.setGain((float) gain);
            midEq.setCutoffFrequency((float) freq);
            eq.setBand(1, midEq);
            dp.setPostEqAllChannelsTo(eq);
        }
        else if (type.equals("treble")) {
            EqBand trebleEq = eq.getBand(2);
            trebleEq.setGain((float) gain);
            trebleEq.setCutoffFrequency((float) freq);
            eq.setBand(2, trebleEq);
            dp.setPostEqAllChannelsTo(eq);
        }
        else {
            System.out.println("adjustEq: invalid eq type");
        }
    }

    /**
     * Returns current tracked EQ
     * @return
     */
    public Map<String, Object> getCurrentEq() {
        Map<String, Object> currentEq = new HashMap<String, Object>();
        currentEq.put(ResponseParameters.bassGain, eq.getBand(0).getGain());
        currentEq.put(ResponseParameters.bassFrequency, eq.getBand(0).getCutoffFrequency());
        currentEq.put(ResponseParameters.midGain, eq.getBand(1).getGain());
        currentEq.put(ResponseParameters.midFrequency, eq.getBand(1).getCutoffFrequency());
        currentEq.put(ResponseParameters.trebleGain, eq.getBand(2).getGain());
        currentEq.put(ResponseParameters.trebleFrequency, eq.getBand(2).getCutoffFrequency());
        return currentEq;
    }

    /**
     * Returns current elapsed time on MediaPlayer
     *
     * Note: This can be done automatically using setElapsedTimeEvent
     *
     * @return
     */
    public Map<String, Object> getElapsedTime() {
        // Is there an event I can subscribe to?
        Map<String, Object> elapsedTime = Utils.timeToDictionary(player.getCurrentPosition());
        return elapsedTime;
    }

    /**
     * Returns total time for the loaded track
     * @return
     */
    public Map<String, Object> getTotalTime() {
        Map<String, Object> totalTime = Utils.timeToDictionary(player.getDuration());
        return totalTime;
    }

    /**
     * Destroys object and resets state.
     *
     * @return
     */
    public Map<String, Object> destroy() {
        stop();
        player.stop();
        player.release();
        dp.release();
        Map<String, Object> response = new HashMap<String, Object>();
        response.put(ResponseParameters.listenerName, listenerName);
        response.put(ResponseParameters.elapsedTimeEventName, elapsedTimeEventName);
        return response;
    }

    /**
     * Handles when a track has completed.
     *
     * @param mediaPlayer
     */
    @Override
    public void onCompletion(MediaPlayer mediaPlayer) {
        try {
            stop();
        }
        catch (Exception ex) {
            Log.e("onCompletion AudioFile", "An error occurred in onCompletion. Exception: " + ex.getLocalizedMessage());
        }
    }

    /**
     * Handles when a track has been initialized
     * @param mediaPlayer
     */
    @Override
    public void onPrepared(MediaPlayer mediaPlayer) {
        try {
            player.seekTo(0);
        }
        catch (Exception ex) {
            Log.e("onPrepared AudioFile", "An error occurred in onPrepared. Exception: " + ex.getLocalizedMessage());
        }
    }

    /**
     * Starts listener for Audio metering.
     *
     * Note: this should only run when being used as it will run continuously in the background
     *       call destroyVisualizerListener to stop process.
     */
    private void initVisualizerListener() {
        if (listenerName.isEmpty()){
            return;
        }
        visualizer = new Visualizer(player.getAudioSessionId());
        visualizer.setScalingMode(Visualizer.SCALING_MODE_AS_PLAYED);
        visualizer.setMeasurementMode(Visualizer.MEASUREMENT_MODE_PEAK_RMS);
        visualizer.setCaptureSize(Visualizer.getCaptureSizeRange()[0]);
        measurementPeakRms = new Visualizer.MeasurementPeakRms();
        visualizer.setDataCaptureListener(new Visualizer.OnDataCaptureListener() {
            @Override
            public void onWaveFormDataCapture(Visualizer vis, byte[] bytes, int i) {
                if(visualizerState) {
                    visualizer.getMeasurementPeakRms(measurementPeakRms);
                    double measurement = (double)measurementPeakRms.mRms;
                    measurement = (measurement / 100) * (1 / currentVolume);
                    double response = measurement < -80 ? -80 : measurement;
                    JSObject data = new JSObject();
                    data.put(ResponseParameters.meterLevel, response);
                    _parent.notifyPluginListeners(listenerName, data);

                    if (!elapsedTimeEventName.isEmpty()) {
                        _parent.notifyPluginListeners(elapsedTimeEventName, Utils.buildResponseData(getElapsedTime()));
                    }
                }

            }

            @Override
            public void onFftDataCapture(Visualizer visualizer, byte[] bytes, int i) {
                Log.i("FFT Byte Array: ", String.valueOf(bytes[0]));
            }
        }, Visualizer.getMaxCaptureRate(), true, false);
        visualizer.setEnabled(true);
        visualizerState = true;
    }

    /**
     * Stop listener for audio metering.
     */
    private void destroyVisualizerListener() {
        if (listenerName.isEmpty()){
            return;
        }
        if (visualizerState) {
            visualizer.setDataCaptureListener(null, 0, false, false);
            visualizer.setEnabled(false);
            visualizerState = false;
            visualizer.release();
        }
        return;
    }
}
